<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Cell Type Identification and Discovery from Single Cell Gene
        Expression Data • SignacX</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Cell Type Identification and Discovery from Single Cell Gene
        Expression Data">
<meta property="og:description" content="An implementation of neural networks trained with flow-sorted gene expression data to classify cellular phenotypes in single cell RNA-sequencing data. See Chamberlain M et al. (2021) &lt;doi:10.1101/2021.02.01.429207&gt; for more details.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">SignacX</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/Crabeating_vignette.html">Mapping homologous gene symbols</a>
    </li>
    <li>
      <a href="articles/SignacFast-Seurat_AMP_RA.html">Benchmarking SignacFast with single cell flow-sorted data</a>
    </li>
    <li>
      <a href="articles/signac-SPRING_Learning.html">Signac and SPRING: Learning CD56 NK cells from multi-modal analysis of CITE-seq PBMCs from 10X Genomics</a>
    </li>
    <li>
      <a href="articles/signac-Seurat_AMP.html">SignacX, Seurat and MASC: Analysis of kidney cells from AMP</a>
    </li>
    <li>
      <a href="articles/signac-Seurat_AMP_RA.html">Benchmarking SignacX and SingleR with synovial flow cytometry data</a>
    </li>
    <li>
      <a href="articles/signac-Seurat_CITE-seq.html">Cell type classification with SignacX: CITE-seq PBMCs from 10X Genomics</a>
    </li>
    <li>
      <a href="articles/signac-Seurat_pbmcs.html">Cell type classification with SignacX: PBMCs from 10X Genomics</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mathewchamberlain/SignacX/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">


<!-- PROJECT LOGO -->
<br><p align="center">
</p>
<h3 align="center">
SignacX 2.2.0
</h3>
<p align="center">
Get the most out of your single cell data. <br><a href="#getting-started"><strong>Explore the docs »</strong></a> <br><br><a href="https://htmlpreview.github.io/?https://github.com/mathewchamberlain/SignacX/master/vignettes/signac-Seurat_pbmcs.html">View Demo</a> · <a href="https://github.com/mathewchamberlain/SignacX/issues">Report Bug</a> · <a href="https://github.com/mathewchamberlain/SignacX/issues">Request Feature</a>
</p>

<!-- TABLE OF CONTENTS -->
<div id="table-of-contents" class="section level2">
<h2 class="hasAnchor">
<a href="#table-of-contents" class="anchor"></a>Table of Contents</h2>
<ul>
<li><a href="#about-the-project">What is SignacX?</a></li>
<li><a href="#data-portal">Data portal</a></li>
<li>
<a href="#getting-started">Getting Started</a>
<ul>
<li><a href="#installation">Installation</a></li>
<li><a href="#quickstart">Quick start</a></li>
</ul>
</li>
<li>
<a href="#usage">Usage</a>
<ul>
<li><a href="#spring">SPRING</a></li>
<li>
<a href="#seurat">Seurat</a>
<ul>
<li><a href="#masc">MASC</a></li>
</ul>
</li>
<li><a href="#non-human-data">Non-human data</a></li>
<li><a href="#genes-of-interest">Genes of interest</a></li>
<li><a href="#learning-from-single-cell-data">Learning from single cell data</a></li>
<li><a href="#fast-signac">Fast Signac</a></li>
</ul>
</li>
<li>
<a href="#benchmarking">Benchmarking</a>
<ul>
<li><a href="#cite-seq">CITE-seq</a></li>
<li><a href="#flow-sorted-synovial-cells">Flow-sorted synovial cells</a></li>
<li><a href="#pbmcs">PBMCs</a></li>
</ul>
</li>
<li><a href="#roadmap">Roadmap</a></li>
<li><a href="#contributing">Contributing</a></li>
<li><a href="#license">License</a></li>
<li><a href="#contact">Contact</a></li>
</ul>
<!-- ABOUT THE PROJECT -->
</div>
<div id="what-is-signacx" class="section level2">
<h2 class="hasAnchor">
<a href="#what-is-signacx" class="anchor"></a>What is SignacX?</h2>
<p>SignacX is software developed and maintained by the Savova lab at Sanofi with a focus on single cell genomics for clinical applications. SignacX classifies the cellular phenotype for each individual cell in single cell RNA-sequencing data using neural networks trained with sorted bulk gene expression data from the <a href="https://bmcgenomics.biomedcentral.com/articles/10.1186/1471-2164-14-632">Human Primary Cell Atlas</a>. In this R implementation, we provide functions and vignettes that demonstrate how to: integrate single cell data (mapping cells from one data set to another), classify non-human data, identify novel cell types, and classify single cell data across many tissues, diseases and technologies. To learn more, check out the pre-print <a href="https://www.biorxiv.org/content/10.1101/2021.02.01.429207v3.full">here</a>.</p>
</div>
<div id="data-portal" class="section level2">
<h2 class="hasAnchor">
<a href="#data-portal" class="anchor"></a>Data portal</h2>
<p>Here, we provide interactive access to data from the <a href="https://www.biorxiv.org/content/10.1101/2021.02.01.429207v3.full">pre-print</a> with <a href="https://kleintools.hms.harvard.edu/tools/spring.html">SPRING Viewer</a>. Just click the “Explore” links below, and search your favorite gene:</p>
<table class="table">
<colgroup>
<col width="14%">
<col width="14%">
<col width="14%">
<col width="14%">
<col width="14%">
<col width="14%">
<col width="14%">
</colgroup>
<thead><tr class="header">
<th>Links</th>
<th>Tissue</th>
<th>Disease</th>
<th>Number of cells</th>
<th>Number of samples</th>
<th>Source</th>
<th>Signac version</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/EXPLORATORY_KIDNEY_IMMUNE_ZONATION/FullDataset_v1">Explore</a></td>
<td>Kidney</td>
<td>Cancer</td>
<td>48,037</td>
<td>47</td>
<td><a href="https://science.sciencemag.org/content/365/6460/1461.abstract">Stewart et al. 2019</a></td>
<td>v2.0.7</td>
</tr>
<tr class="even">
<td><a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/AMP_Phase1_SLE_Apr2019/FullDataset_v1">Explore</a></td>
<td>Kidney and urine</td>
<td>Lupus nephritis and healthy</td>
<td>5,886</td>
<td>39</td>
<td><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6726437/">Arazi et al. 2019</a></td>
<td>v2.0.7</td>
</tr>
<tr class="odd">
<td><a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/Immunity_LungCancer_lung/FullDataset_v1">Explore</a></td>
<td>Lung</td>
<td>Cancer</td>
<td>42,844</td>
<td>18</td>
<td><a href="https://pubmed.ncbi.nlm.nih.gov/30979687/">Zilionis et al. 2020</a></td>
<td>v2.0.7</td>
</tr>
<tr class="even">
<td><a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/EXPLORATORY_FIBROTICLUNGARIZONA/FullDataset_v1">Explore</a></td>
<td>Lung</td>
<td>Fibrosis</td>
<td>96,461</td>
<td>31</td>
<td><a href="https://advances.sciencemag.org/content/6/28/eaba1972">Habermann et al. 2020</a></td>
<td>v2.0.7</td>
</tr>
<tr class="odd">
<td><a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/EXPLORATORY_FIBROTICLUNGCHICAGO/FullDataset_v1">Explore</a></td>
<td>Lung</td>
<td>Fibrosis</td>
<td>109,421</td>
<td>16</td>
<td><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6580683/">Reyfman et al. 2019</a></td>
<td>v2.0.7</td>
</tr>
<tr class="even">
<td><a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/Monkey_Sample3003/FullDataset_v1">Explore</a></td>
<td>Monkey PBMCs</td>
<td>Healthy</td>
<td>5,491</td>
<td>1</td>
<td><a href="https://www.biorxiv.org/content/10.1101/2021.02.01.429207v3.full">Chamberlain et al. 2021</a></td>
<td>v2.0.7</td>
</tr>
<tr class="odd">
<td><a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/Monkey_Sample3004/FullDataset_v1">Explore</a></td>
<td>Monkey PBMCs</td>
<td>Healthy</td>
<td>5,220</td>
<td>1</td>
<td><a href="https://www.biorxiv.org/content/10.1101/2021.02.01.429207v3.full">Chamberlain et al. 2021</a></td>
<td>v2.0.7</td>
</tr>
<tr class="even">
<td><a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/Monkey_Tcells/FullDataset_v1">Explore</a></td>
<td>Monkey T cells</td>
<td>Healthy</td>
<td>5,496</td>
<td>1</td>
<td><a href="https://www.biorxiv.org/content/10.1101/2021.02.01.429207v3.full">Chamberlain et al. 2021</a></td>
<td>v2.0.7</td>
</tr>
<tr class="odd">
<td><a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/Immunity_LungCancer_blood/FullDataset_v1">Explore</a></td>
<td>PBMCs</td>
<td>Cancer</td>
<td>14,048</td>
<td>8</td>
<td><a href="https://pubmed.ncbi.nlm.nih.gov/30979687/">Zilionis et al. 2020</a></td>
<td>v2.0.7</td>
</tr>
<tr class="even">
<td><a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/CITESEQ_EXPLORATORY_CITESEQ_5K_PBMCS/FullDataset_v1_protein">Explore</a></td>
<td>PBMCs</td>
<td>Healthy</td>
<td>7,902</td>
<td>1</td>
<td><a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/pbmc_10k_protein_v3">10X Genomics</a></td>
<td>v2.0.7</td>
</tr>
<tr class="odd">
<td><a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/PBMCs_5k_10X/FullDataset_v1">Explore</a></td>
<td>PBMCs</td>
<td>Healthy</td>
<td>4,784</td>
<td>1</td>
<td><a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.2/5k_pbmc_v3">10X Genomics</a></td>
<td>v2.0.7</td>
</tr>
<tr class="even">
<td><a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/EXPLORATORY_AD_FIBROBLASTS_EMMA_GUTTMAN/FullDataset_v1">Explore</a></td>
<td>Skin</td>
<td>Atopic dermatitis</td>
<td>36,690</td>
<td>17</td>
<td><a href="https://pubmed.ncbi.nlm.nih.gov/32035984/">He et al. 2020</a></td>
<td>v2.0.7</td>
</tr>
<tr class="odd">
<td><a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/AMP_Phase1_RA_Apr2019/FullDataset_v1">Explore</a></td>
<td>Synovium</td>
<td>Rheumatoid arthritis and osteoarthritis</td>
<td>8,920</td>
<td>26</td>
<td><a href="https://www.nature.com/articles/s41590-019-0378-1">Zhang et. al 2019</a></td>
<td>v2.0.7</td>
</tr>
</tbody>
</table>
<p>Note:</p>
<ul>
<li>Cell type annotations are provided at four levels (immune, celltypes, cellstates and novel celltypes).</li>
<li>When available, we also provided information about sample covariates (i.e., disease, age, gender, FACs etc.).</li>
<li>Cell type annotations for all 13 data sets were generated with the Signac function with the default settings without changing any settings or parameters.</li>
</ul>
<p>Special thanks to Allon Klein’s lab (particularly Caleb Weinreb and Sam Wolock) for hosting the data.</p>
<!-- GETTING STARTED -->
</div>
<div id="getting-started" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting Started</h2>
<p>To install SignacX in R, simply do:</p>
<div id="installation" class="section level3">
<h3 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h3>
<div id="install-from-cran" class="section level4">
<h4 class="hasAnchor">
<a href="#install-from-cran" class="anchor"></a>Install from CRAN</h4>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">'SignacX'</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mathewchamberlain/SignacX">SignacX</a></span><span class="op">)</span></code></pre></div>
</div>
<div id="install-development-version-of-signacx" class="section level4">
<h4 class="hasAnchor">
<a href="#install-development-version-of-signacx" class="anchor"></a>Install development version of SignacX</h4>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"mathewchamberlain/SignacX"</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mathewchamberlain/SignacX">SignacX</a></span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="quick-start" class="section level3">
<h3 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick start</h3>
<p>The main functions in Signac are:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load the library</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mathewchamberlain/SignacX">SignacX</a></span><span class="op">)</span>

<span class="co"># Generate initial labels</span>
<span class="va">labels</span> <span class="op">=</span> <span class="fu"><a href="reference/Signac.html">Signac</a></span><span class="op">(</span>E <span class="op">=</span> <span class="va">your_data_here</span><span class="op">)</span>

<span class="co"># Get cell type labels</span>
<span class="va">celltypes</span> <span class="op">=</span> <span class="fu"><a href="reference/GenerateLabels.html">GenerateLabels</a></span><span class="op">(</span><span class="va">labels</span>, E <span class="op">=</span> <span class="va">your_data_here</span><span class="op">)</span></code></pre></div>
<p>Sometimes we don’t have time to run Signac, and need a quick solution. Although Signac scales fine with large data sets (&gt;300,000 cells), we developed SignacFast to quickly classify single cell data:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># generate labels with pre-trained model</span>
<span class="va">labels_fast</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/SignacFast.html">SignacFast</a></span><span class="op">(</span>E <span class="op">=</span> <span class="va">your_data_here</span>, num.cores <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>
<span class="va">celltypes_fast</span> <span class="op">=</span> <span class="fu"><a href="reference/GenerateLabels.html">GenerateLabels</a></span><span class="op">(</span><span class="va">labels_fast</span>, E <span class="op">=</span> <span class="va">your_data_here</span><span class="op">)</span></code></pre></div>
<!-- USAGE EXAMPLES -->
</div>
</div>
<div id="usage" class="section level2">
<h2 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h2>
<p>To make life easier, SignacX was integrated with <a href="https://satijalab.org/seurat/">Seurat</a> (versions 3 and 4), and with <a href="https://pubmed.ncbi.nlm.nih.gov/29228172/">SPRING</a>. We provide a few vignettes:</p>
<div id="spring" class="section level3">
<h3 class="hasAnchor">
<a href="#spring" class="anchor"></a>SPRING</h3>
<p>In the <a href="https://www.biorxiv.org/content/10.1101/2021.02.01.429207v3.full">pre-print</a>, we often used Signac integrated with <a href="https://pubmed.ncbi.nlm.nih.gov/29228172/">SPRING</a>. To reproduce our findings and to generate new results with SPRING, please visit the SPRING repository which has <a href="https://github.com/AllonKleinLab/SPRING_dev">example notebooks and installation instructions</a>, particularly for <a href="https://github.com/AllonKleinLab/SPRING_dev/blob/master/data_prep/spring_notebook_10X_CITEseq.ipynb">processing CITE-seq and scRNA-seq data from 10X Genomics</a>. Briefly, Signac is integrated seamlessly with the output files of SPRING in R, requiring only a few functions:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># dir points to the "FullDataset_v1" directory generated by the SPRING Jupyter notebook</span>
<span class="va">dir</span> <span class="op">=</span> <span class="st">"./FullDataset_v1"</span> 

<span class="co"># load the expression data</span>
<span class="va">E</span> <span class="op">=</span> <span class="fu"><a href="reference/CID.LoadData.html">CID.LoadData</a></span><span class="op">(</span><span class="va">dir</span><span class="op">)</span>

<span class="co"># generate cellular phenotype labels</span>
<span class="va">labels</span> <span class="op">=</span> <span class="fu"><a href="reference/Signac.html">Signac</a></span><span class="op">(</span><span class="va">E</span>, spring.dir <span class="op">=</span> <span class="va">dir</span><span class="op">)</span>
<span class="va">celltypes</span> <span class="op">=</span> <span class="fu"><a href="reference/GenerateLabels.html">GenerateLabels</a></span><span class="op">(</span><span class="va">labels</span>, E <span class="op">=</span> <span class="va">E</span>, spring.dir <span class="op">=</span> <span class="va">dir</span><span class="op">)</span>

<span class="co"># write cell types and Louvain clusters to SPRING</span>
<span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/CID.writeJSON.html">CID.writeJSON</a></span><span class="op">(</span><span class="va">celltypes</span>, spring.dir <span class="op">=</span> <span class="va">dir</span><span class="op">)</span></code></pre></div>
<p>After running the above functions, cellular phenotypes and Louvain clusters are ready to be visualized with SPRING Viewer, which can be setup locally as described <a href="https://github.com/AllonKleinLab/SPRING_dev">here</a>.</p>
</div>
<div id="seurat" class="section level3">
<h3 class="hasAnchor">
<a href="#seurat" class="anchor"></a>Seurat</h3>
<p>Another way to use Signac is with Seurat. <a href="https://htmlpreview.github.io/?https://github.com/mathewchamberlain/SignacX/master/vignettes/signac-Seurat_CITE-seq.html">In this vignette</a>, we performed multi-modal analysis of CITE-seq PBMCs from 10X Genomics using Signac integrated with Seurat.</p>
<p>Note:</p>
<ul>
<li>This same data set was also processed using SPRING <a href="https://github.com/AllonKleinLab/SPRING_dev/blob/master/data_prep/spring_notebook_10X_CITEseq.ipynb">in this notebook</a>, and subsequently classified with Signac, which was used to generate SPRING layouts for these data in the <a href="https://www.biorxiv.org/content/10.1101/2021.02.01.429207v3.full">pre-print</a> (Figures 2-4), which is available for interactive exploration <a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/CITESEQ_EXPLORATORY_CITESEQ_5K_PBMCS/FullDataset_v1_protein">here</a>.</li>
</ul>
<div id="masc" class="section level4">
<h4 class="hasAnchor">
<a href="#masc" class="anchor"></a>MASC</h4>
<p>Sometimes, we have single cell genomics data with disease information, and we want to know which cellular phenotypes are enriched for disease. <a href="https://htmlpreview.github.io/?https://github.com/mathewchamberlain/SignacX/master/vignettes/signac-Seurat_AMP.html">In this vignette</a>, we applied Signac to classify cellular phenotypes in healthy and lupus nephritis kidney cells, and then we used <a href="https://pubmed.ncbi.nlm.nih.gov/30333237/">MASC</a> to identify which cellular phenotypes were disease-enriched.</p>
<p>Note:</p>
<ul>
<li>MASC typically requires equal numbers of cells and samples between case and control: an unequal number might skew the clustering of cells towards one sample (i.e., a “batch effect”), which could cause spurious disease enrichment in the mixed effect model. Since Signac classifies each cell independently (without using clusters), Signac annotations can be used with MASC without a priori balancing samples or cells, unlike cluster-based annotation methods.</li>
</ul>
</div>
</div>
<div id="non-human-data" class="section level3">
<h3 class="hasAnchor">
<a href="#non-human-data" class="anchor"></a>Non-human data</h3>
<p>In Supplemental Figure 8 of the <a href="https://www.biorxiv.org/content/10.1101/2021.02.01.429207v3.full">pre-print</a>, we classified single cell data for a model organism (cynomolgus monkey) for which flow-sorted datasets were generally lacking without any additional species-specific training. Instead, we mapped homologous genes from the <em>Macaca fascicularis</em> genome to the human genome in the single cell data, and then performed cell type classification with Signac. We demonstrate how we mapped the gene symbols <a href="https://htmlpreview.github.io/?https://github.com/mathewchamberlain/SignacX/master/vignettes/Crabeating_vignette.html">here</a>.</p>
<p>Note:</p>
<ul>
<li>This code can be used for to identify homologous genes between any two species.</li>
<li>Monkey data used in Supplemental Figure 8 are available for interactive exploration in the table listed above.</li>
</ul>
</div>
<div id="genes-of-interest" class="section level3">
<h3 class="hasAnchor">
<a href="#genes-of-interest" class="anchor"></a>Genes of interest</h3>
<p>In Figure 6 of the <a href="https://www.biorxiv.org/content/10.1101/2021.02.01.429207v3.full">pre-print</a>, we compiled data from three source (<a href="https://www.cellphonedb.org/">CellPhoneDB</a>, <a href="https://www.ebi.ac.uk/gwas/">GWAS catalog</a> and <a href="https://www.nature.com/articles/s41588-019-0456-1">Fang et al. 2020</a>) to find genes of immunological / pharmacological interest. These genes and their annotations can be accessed internally from within Signac:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># See ?Genes_Of_Interest</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"Genes_Of_Interest"</span><span class="op">)</span></code></pre></div>
</div>
<div id="learning-from-single-cell-data" class="section level3">
<h3 class="hasAnchor">
<a href="#learning-from-single-cell-data" class="anchor"></a>Learning from single cell data</h3>
<p>In Figure 4 of the <a href="https://www.biorxiv.org/content/10.1101/2021.02.01.429207v3.full">pre-print</a>, we demonstrated that Signac mapped cell type labels from one single cell data set to another; learning CD56<sup>bright</sup> NK cells from CITE-seq data. <a href="https://htmlpreview.github.io/?https://github.com/mathewchamberlain/SignacX/master/vignettes/signac-SPRING_Learning.html">Here, we provide a vignette</a> for reproducing this analysis, which can be used to map cell populations (or clusters of cells) from one data set to another. We also provide interactive access to the single cell data that were annotated with the CD56<sup>bright</sup> NK cell-model (Note: the CD56<sup>bright</sup> NK cells appear in the “CellStates” annotation layer as red cells).</p>
<table class="table">
<colgroup>
<col width="14%">
<col width="14%">
<col width="14%">
<col width="14%">
<col width="14%">
<col width="14%">
<col width="14%">
</colgroup>
<thead><tr class="header">
<th>Links</th>
<th>Tissue</th>
<th>Disease</th>
<th>Number of cells</th>
<th>Number of samples</th>
<th>Source</th>
<th>Signac version</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/EXPLORATORY_KIDNEY_IMMUNE_ZONATION/FullDataset_v1_Learned">Explore</a></td>
<td>Kidney</td>
<td>Cancer</td>
<td>48,037</td>
<td>47</td>
<td><a href="https://science.sciencemag.org/content/365/6460/1461.abstract">Stewart et al. 2019</a></td>
<td>v2.0.7 + CD56<sup>bright</sup> NK</td>
</tr>
<tr class="even">
<td><a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/AMP_Phase1_SLE_Apr2019/FullDataset_v1_Learned">Explore</a></td>
<td>Kidney and urine</td>
<td>Lupus nephritis and healthy</td>
<td>5,886</td>
<td>39</td>
<td><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6726437/">Arazi et al. 2019</a></td>
<td>v2.0.7 + CD56<sup>bright</sup> NK</td>
</tr>
<tr class="odd">
<td><a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/Immunity_LungCancer_lung/FullDataset_v1_Learned">Explore</a></td>
<td>Lung</td>
<td>Cancer</td>
<td>42,844</td>
<td>18</td>
<td><a href="https://pubmed.ncbi.nlm.nih.gov/30979687/">Zilionis et al. 2020</a></td>
<td>v2.0.7 + CD56<sup>bright</sup> NK</td>
</tr>
<tr class="even">
<td><a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/EXPLORATORY_FIBROTICLUNGARIZONA/FullDataset_v1_Learned">Explore</a></td>
<td>Lung</td>
<td>Fibrosis</td>
<td>96,461</td>
<td>31</td>
<td><a href="https://advances.sciencemag.org/content/6/28/eaba1972">Habermann et al. 2020</a></td>
<td>v2.0.7 + CD56<sup>bright</sup> NK</td>
</tr>
<tr class="odd">
<td><a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/EXPLORATORY_FIBROTICLUNGCHICAGO/FullDataset_v1_Learned">Explore</a></td>
<td>Lung</td>
<td>Fibrosis</td>
<td>109,421</td>
<td>16</td>
<td><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6580683/">Reyfman et al. 2019</a></td>
<td>v2.0.7 + CD56<sup>bright</sup> NK</td>
</tr>
<tr class="even">
<td><a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/Monkey_Sample3003/FullDataset_v1_Learned">Explore</a></td>
<td>Monkey PBMCs</td>
<td>Healthy</td>
<td>5,491</td>
<td>1</td>
<td><a href="https://www.biorxiv.org/content/10.1101/2021.02.01.429207v3.full">Chamberlain et al. 2021</a></td>
<td>v2.0.7 + CD56<sup>bright</sup> NK</td>
</tr>
<tr class="odd">
<td><a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/Monkey_Sample3004/FullDataset_v1_Learned">Explore</a></td>
<td>Monkey PBMCs</td>
<td>Healthy</td>
<td>5,220</td>
<td>1</td>
<td><a href="https://www.biorxiv.org/content/10.1101/2021.02.01.429207v3.full">Chamberlain et al. 2021</a></td>
<td>v2.0.7 + CD56<sup>bright</sup> NK</td>
</tr>
<tr class="even">
<td><a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/Monkey_Tcells/FullDataset_v1_Learned">Explore</a></td>
<td>Monkey T cells</td>
<td>Healthy</td>
<td>5,496</td>
<td>1</td>
<td><a href="https://www.biorxiv.org/content/10.1101/2021.02.01.429207v3.full">Chamberlain et al. 2021</a></td>
<td>v2.0.7 + CD56<sup>bright</sup> NK</td>
</tr>
<tr class="odd">
<td><a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/Immunity_LungCancer_blood/FullDataset_v1_Learned">Explore</a></td>
<td>PBMCs</td>
<td>Cancer</td>
<td>14,048</td>
<td>8</td>
<td><a href="https://pubmed.ncbi.nlm.nih.gov/30979687/">Zilionis et al. 2020</a></td>
<td>v2.0.7 + CD56<sup>bright</sup> NK</td>
</tr>
<tr class="even">
<td><a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/PBMCs_5k_10X/FullDataset_v1_Learned">Explore</a></td>
<td>PBMCs</td>
<td>Healthy</td>
<td>4,784</td>
<td>1</td>
<td><a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.2/5k_pbmc_v3">10X Genomics</a></td>
<td>v2.0.7 + CD56<sup>bright</sup> NK</td>
</tr>
<tr class="odd">
<td><a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/EXPLORATORY_AD_FIBROBLASTS_EMMA_GUTTMAN/FullDataset_v1_Learned">Explore</a></td>
<td>Skin</td>
<td>Atopic dermatitis</td>
<td>36,690</td>
<td>17</td>
<td><a href="https://pubmed.ncbi.nlm.nih.gov/32035984/">He et al. 2020</a></td>
<td>v2.0.7 + CD56<sup>bright</sup> NK</td>
</tr>
<tr class="even">
<td><a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/AMP_Phase1_RA_Apr2019/FullDataset_v1">Explore</a></td>
<td>Synovium</td>
<td>Rheumatoid arthritis and osteoarthritis</td>
<td>8,920</td>
<td>26</td>
<td><a href="https://www.nature.com/articles/s41590-019-0378-1">Zhang et. al 2019</a></td>
<td>v2.0.7 + CD56<sup>bright</sup> NK</td>
</tr>
</tbody>
</table>
</div>
<div id="fast-signac" class="section level3">
<h3 class="hasAnchor">
<a href="#fast-signac" class="anchor"></a>Fast Signac</h3>
<p>Sometimes we don’t have time to run Signac and need a faster solution. Although Signac scales fine with large data sets (&gt;300,000 cells) and even for large data, typically takes less than an hour, we developed SignacFast to quickly classify single cell data:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># generate labels with pre-trained model</span>
<span class="va">labels_fast</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/SignacFast.html">SignacFast</a></span><span class="op">(</span>E <span class="op">=</span> <span class="va">your_data_here</span>, num.cores <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>
<span class="va">celltypes_fast</span> <span class="op">=</span> <span class="fu"><a href="reference/GenerateLabels.html">GenerateLabels</a></span><span class="op">(</span><span class="va">labels_fast</span>, E <span class="op">=</span> <span class="va">your_data_here</span><span class="op">)</span></code></pre></div>
<p>Unlike Signac, SignacFast uses a pre-trained ensemble of neural network models generated from the HPCA reference data, speeding classsification time ~5-10x fold. These models were generated from the HPCA training data like so:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load pre-trained neural network ensemble model</span>
<span class="va">ref</span> <span class="op">=</span> <span class="fu"><a href="reference/GetTrainingData_HPCA.html">GetTrainingData_HPCA</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># generate models</span>
<span class="va">Models_HPCA</span> <span class="op">=</span> <span class="fu"><a href="reference/ModelGenerator.html">ModelGenerator</a></span><span class="op">(</span>R <span class="op">=</span> <span class="va">training_HPCA</span>, N <span class="op">=</span> <span class="fl">100</span>, num.cores <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></code></pre></div>
<p>The “Models_HPCA” are accessed from within the R package:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load pre-trained neural network ensemble model</span>
<span class="va">Models</span> <span class="op">=</span> <span class="fu"><a href="reference/GetModels_HPCA.html">GetModels_HPCA</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>We demonstrate how to use SignacFast in this <a href="https://htmlpreview.github.io/?https://github.com/mathewchamberlain/SignacX/master/vignettes/SignacFast-Seurat_AMP_RA.html">vignette</a>, which shows that the results are broadly consistent with running Signac.</p>
<p>Note:</p>
<ul>
<li>For proper use; if the concern is only major cell types (i.e., TNK and MPh cells), then SignacFast is a fine alternative to Signac.</li>
</ul>
<!-- BENCHMARKING -->
</div>
</div>
<div id="benchmarking" class="section level2">
<h2 class="hasAnchor">
<a href="#benchmarking" class="anchor"></a>Benchmarking</h2>
<div id="cite-seq" class="section level3">
<h3 class="hasAnchor">
<a href="#cite-seq" class="anchor"></a>CITE-seq</h3>
<p>In Figure 2-3 of the <a href="https://www.biorxiv.org/content/10.1101/2021.02.01.429207v3.full">pre-print</a>, we validated Signac with CITE-seq PBMCs. Here, we reproduced that analysis with SPRING (<a href="https://github.com/AllonKleinLab/SPRING_dev">in this vignette</a>; as was performed in the pre-print) and additionally with Seurat (<a href="https://htmlpreview.github.io/?https://github.com/mathewchamberlain/SignacX/master/vignettes/signac-Seurat_CITE-seq.html">in this vignette</a>), and provide interactive access to the data <a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/CITESEQ_EXPLORATORY_CITESEQ_5K_PBMCS/FullDataset_v1_protein">here</a>.</p>
</div>
<div id="flow-sorted-synovial-cells" class="section level3">
<h3 class="hasAnchor">
<a href="#flow-sorted-synovial-cells" class="anchor"></a>Flow-sorted synovial cells</h3>
<p>In Figure 3 of the <a href="https://www.biorxiv.org/content/10.1101/2021.02.01.429207v3.full">pre-print</a>, we validated Signac with flow cytometry and compared Signac to SingleR. We reproduced that analysis using Seurat <a href="https://htmlpreview.github.io/?https://github.com/mathewchamberlain/SignacX/master/vignettes/signac-Seurat_AMP_RA.html">in this vignette</a>, and provide interactive access to the data <a href="https://kleintools.hms.harvard.edu/tools/springViewer_1_6_dev.html?client_datasets/AMP_Phase1_RA_Apr2019/FullDataset_v1">here</a>.</p>
</div>
<div id="pbmcs" class="section level3">
<h3 class="hasAnchor">
<a href="#pbmcs" class="anchor"></a>PBMCs</h3>
<p>In Table 1 of the <a href="https://www.biorxiv.org/content/10.1101/2021.02.01.429207v3.full">pre-print</a>, we benchmarked Signac across seven different technologies: CEL-seq, Drop-Seq, inDrop, 10X (v2), 10X (v3), Seq-Well and Smart-Seq2; this analysis was reproduced <a href="https://htmlpreview.github.io/?https://github.com/mathewchamberlain/SignacX/master/vignettes/PBMC_bench.html">here</a>.</p>
<!-- ROADMAP -->
</div>
</div>
<div id="roadmap" class="section level2">
<h2 class="hasAnchor">
<a href="#roadmap" class="anchor"></a>Roadmap</h2>
<p>See the <a href="https://github.com/mathewchamberlain/SignacX/issues">open issues</a> for a list of proposed features (and known issues).</p>
<!-- CONTRIBUTING -->
</div>
<div id="contributing" class="section level2">
<h2 class="hasAnchor">
<a href="#contributing" class="anchor"></a>Contributing</h2>
<p>Any contributions you make are <strong>greatly appreciated</strong>.</p>
<ol>
<li>Fork the Project</li>
<li>Create your Feature Branch (<code>git checkout -b feature/AmazingFeature</code>)</li>
<li>Commit your Changes (<code>git commit -m 'Add some AmazingFeature'</code>)</li>
<li>Push to the Branch (<code>git push origin feature/AmazingFeature</code>)</li>
<li>Open a Pull Request</li>
</ol>
<p>You can also open a pull request to commit to the master branch.</p>
<!-- LICENSE -->
</div>
<div id="license" class="section level2">
<h2 class="hasAnchor">
<a href="#license" class="anchor"></a>License</h2>
<p>Distributed under the GPL v3.0 License. See <code>LICENSE</code> for more information.</p>
<!-- CONTACT -->
</div>
<div id="contact" class="section level2">
<h2 class="hasAnchor">
<a href="#contact" class="anchor"></a>Contact</h2>
<p>Mathew Chamberlain - <a href="mailto:mathew.chamberlain@sanofi.com" class="email">mathew.chamberlain@sanofi.com</a></p>
<p>Project Link: <a href="https://github.com/mathewchamberlain/SignacX">https://github.com/mathewchamberlain/SignacX</a></p>
<!-- MARKDOWN LINKS & IMAGES -->
<!-- https://www.markdownguide.org/basic-syntax/#reference-style-links -->
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from CRAN at <br><a href="https://cloud.r-project.org/package=SignacX">https://​cloud.r-project.org/​package=SignacX</a>
</li>
<li>Browse source code at <br><a href="https://github.com/mathewchamberlain/SignacX/">https://​github.com/​mathewchamberlain/​SignacX/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/mathewchamberlain/SignacX/issues">https://​github.com/​mathewchamberlain/​SignacX/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing SignacX</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Mathew Chamberlain <br><small class="roles"> Author, maintainer </small>  </li>
<li>Virginia Savova <br><small class="roles"> Author </small>  </li>
<li>Richa Hanamsagar <br><small class="roles"> Author </small>  </li>
<li>Frank Nestle <br><small class="roles"> Author </small>  </li>
<li>Emanuele de Rinaldis <br><small class="roles"> Author </small>  </li>
<li>Sanofi US <br><small class="roles"> Funder </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/mathewchamberlain/SignacX/graphs/contributors"><img src="https://img.shields.io/github/contributors/mathewchamberlain/SignacX.svg?style=flat-square" alt="Contributors"></a></li>
<li><a href="https://github.com/mathewchamberlain/SignacX/network/members"><img src="https://img.shields.io/github/forks/mathewchamberlain/SignacX.svg?style=flat-square" alt="Forks"></a></li>
<li><a href="https://github.com/mathewchamberlain/SignacX/stargazers"><img src="https://img.shields.io/github/stars/mathewchamberlain/SignacX.svg?style=flat-square" alt="Stargazers"></a></li>
<li><a href="https://github.com/mathewchamberlain/SignacX/issues"><img src="https://img.shields.io/github/issues/mathewchamberlain/SignacX.svg?style=flat-square" alt="Issues"></a></li>
<li><a href="https://choosealicense.com/licenses/gpl-3.0/"><img src="https://img.shields.io/github/license/mathewchamberlain/SignacX.svg?style=flat-square" alt="GPL License"></a></li>
<li><a href="https://cran.r-project.org/package=SignacX"><img src="https://www.r-pkg.org/badges/version/SignacX" alt="CRAN Version"></a></li>
<li><a href="https://cran.r-project.org/package=SignacX"><img src="https://cranlogs.r-pkg.org/badges/SignacX" alt="CRAN Downloads"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Mathew Chamberlain, Virginia Savova, Richa Hanamsagar, Frank Nestle, Emanuele de Rinaldis, Sanofi US.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
