<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Mapping homologous gene symbols â€¢ SignacX</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Mapping homologous gene symbols">
<meta property="og:description" content="SignacX">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SignacX</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/install.html">Install</a>
</li>
<li>
  <a href="../articles/dataportal.html">Data portal</a>
</li>
<li>
  <a href="../articles/quickstart.html">Quick start</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Introductory Vignettes</li>
    <li>
      <a href="../articles/signac-Seurat_pbmcs.html">PBMCs 1K tutorial</a>
    </li>
    <li>
      <a href="../articles/springintegration.html">Integrating SignacX with SPRING and Seurat</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Advanced use</li>
    <li>
      <a href="../articles/signac-SPRING_Learning.html">Learning from single cell data</a>
    </li>
    <li>
      <a href="../articles/Crabeating_vignette.html">Classifying non-human single cell data</a>
    </li>
    <li>
      <a href="../articles/signac-Seurat_AMP.html">Identifying disease-enriched cell types with SignacX</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Benchmarking</li>
    <li>
      <a href="../articles/signac-Seurat_AMP_RA.html">Flow cytometry synovium: SingleR vs. SignacX</a>
    </li>
    <li>
      <a href="../articles/SignacFast-Seurat_AMP_RA.html">Flow cytometry synovium: FastSignac vs Signac</a>
    </li>
    <li>
      <a href="../articles/signac-Seurat_CITE-seq.html">CITE-seq PBMCs</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Other</li>
    <li>
      <a href="../articles/genesofinterest.html">Genes of immune and pharmacological interest</a>
    </li>
  </ul>
</li>
<li>
  <a href="https://github.com/mathewchamberlain/SignacX/discussions">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mathewchamberlain/SignacX/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Crabeating_vignette_files/header-attrs-2.7/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Mapping homologous gene symbols</h1>
            
            <h4 class="date">Compiled 2021-03-01</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/mathewchamberlain/SignacX/blob/master/vignettes/Crabeating_vignette.Rmd"><code>vignettes/Crabeating_vignette.Rmd</code></a></small>
      <div class="hidden name"><code>Crabeating_vignette.Rmd</code></div>

    </div>

    
    
<p>In Supplemental Figure 8 of the <a href="https://www.biorxiv.org/content/10.1101/2021.02.01.429207v3.full">pre-print</a>, we classified single cell data for a model organism (cynomolgus monkey) for which flow-sorted datasets were generally lacking without any additional species-specific training. Instead, we mapped homologous genes from the <em>Macaca fascicularis</em> genome to the human genome in the single cell data, and then performed cell type classification with Signac. We demonstrate how we mapped the gene symbols in this vignette.</p>
<p>Note: * This code can be used for to identify homologous genes between any two species. * Monkey data used in Supplemental Figure 8 are available for interactive exploration in the table listed above.</p>
<p>This vignette shows how to map homologous gene symbols from Macaca fascicularis to the human genome.</p>
<div id="load-the-essential-packages" class="section level1">
<h1 class="hasAnchor">
<a href="#load-the-essential-packages" class="anchor"></a>Load the essential packages</h1>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va">biomaRt</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va">tidyverse</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mathewchamberlain/SignacX">SignacX</a></span><span class="op">)</span></code></pre></div>
<p>After mapping the reads to the Macaca fasccicularis genome, we load the genes, which were generated from the output of the <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">cellranger pipeline</a> from 10X Genomics.</p>
</div>
<div id="load-macaca-fascicularis-genes" class="section level1">
<h1 class="hasAnchor">
<a href="#load-macaca-fascicularis-genes" class="anchor"></a>Load Macaca fascicularis genes</h1>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">features.tsv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="st">"fls/features.tsv.gz"</span>, header <span class="op">=</span> <span class="cn">FALSE</span>, stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">features.tsv</span><span class="op">)</span></code></pre></div>
<pre><code>##                   V1                 V2              V3
## 1 ENSMFAG00000044637              PGBD2 Gene Expression
## 2 ENSMFAG00000039056             ZNF692 Gene Expression
## 3 ENSMFAG00000030010             ZNF672 Gene Expression
## 4 ENSMFAG00000002737            SH3BP5L Gene Expression
## 5 ENSMFAG00000000508              LYPD8 Gene Expression
## 6 ENSMFAG00000040572 ENSMFAG00000040572 Gene Expression</code></pre>
</div>
<div id="map-homologous-genes" class="section level1">
<h1 class="hasAnchor">
<a href="#map-homologous-genes" class="anchor"></a>Map homologous genes</h1>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get human and cyno gene symbols</span>
<span class="va">human.R95</span> <span class="op">&lt;-</span> <span class="fu">useMart</span><span class="op">(</span>host <span class="op">=</span> <span class="st">"jan2019.archive.ensembl.org"</span>, biomart <span class="op">=</span> <span class="st">"ENSEMBL_MART_ENSEMBL"</span>, dataset <span class="op">=</span> <span class="st">"hsapiens_gene_ensembl"</span><span class="op">)</span>
<span class="va">cyno.R95</span> <span class="op">&lt;-</span> <span class="fu">useMart</span><span class="op">(</span>host <span class="op">=</span> <span class="st">"jan2019.archive.ensembl.org"</span>, biomart <span class="op">=</span> <span class="st">"ENSEMBL_MART_ENSEMBL"</span>, dataset <span class="op">=</span> <span class="st">"mfascicularis_gene_ensembl"</span><span class="op">)</span>

<span class="co"># values = listeENSID: list of cynomolgus ensembl IDs to be retrieved.</span>
<span class="va">listeENSID</span> <span class="op">=</span> <span class="va">features.tsv</span><span class="op">$</span><span class="va">V1</span>
<span class="va">orthologs</span> <span class="op">&lt;-</span> <span class="fu">getLDS</span><span class="op">(</span>attributes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ensembl_gene_id"</span>, <span class="st">"external_gene_name"</span><span class="op">)</span>, filters <span class="op">=</span> <span class="st">"ensembl_gene_id"</span>, 
    values <span class="op">=</span> <span class="va">listeENSID</span>, mart <span class="op">=</span> <span class="va">cyno.R95</span>, attributesL <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"hgnc_symbol"</span>, <span class="st">"ensembl_gene_id"</span><span class="op">)</span>, martL <span class="op">=</span> <span class="va">human.R95</span><span class="op">)</span>
<span class="va">orthologs</span> <span class="op">&lt;-</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="va">orthologs</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">orthologs</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GeneID"</span>, <span class="st">"cynoSymbol"</span>, <span class="st">"HumanSymbol"</span>, <span class="st">"HumanGeneID"</span><span class="op">)</span>

<span class="co"># keep only 1:1 orthologs</span>
<span class="va">one2one</span> <span class="op">&lt;-</span> <span class="va">orthologs</span> <span class="op">%&gt;%</span> <span class="fu">group_by</span><span class="op">(</span><span class="va">GeneID</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">summarise</span><span class="op">(</span><span class="fu">n</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">`n()`</span> <span class="op">&lt;=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">GeneID</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu">pull</span><span class="op">(</span><span class="op">)</span>
<span class="va">orthologs</span> <span class="op">&lt;-</span> <span class="va">orthologs</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">GeneID</span> <span class="op">%in%</span> <span class="va">one2one</span><span class="op">)</span>

<span class="co"># replace empty HumanSymbol (where there isn't a gene name for a homologous gene) with NA</span>
<span class="va">orthologs</span> <span class="op">&lt;-</span> <span class="va">orthologs</span> <span class="op">%&gt;%</span> <span class="fu">mutate</span><span class="op">(</span>HumanSymbol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/replace.html">replace</a></span><span class="op">(</span><span class="va">HumanSymbol</span>, <span class="va">HumanSymbol</span> <span class="op">==</span> <span class="st">""</span>, <span class="cn">NA</span><span class="op">)</span><span class="op">)</span>
<span class="va">orthologs</span> <span class="op">&lt;-</span> <span class="va">orthologs</span> <span class="op">%&gt;%</span> <span class="fu">mutate</span><span class="op">(</span>cynoSymbol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/replace.html">replace</a></span><span class="op">(</span><span class="va">cynoSymbol</span>, <span class="va">cynoSymbol</span> <span class="op">==</span> <span class="st">""</span>, <span class="cn">NA</span><span class="op">)</span><span class="op">)</span>

<span class="va">idx</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="va">listeENSID</span>, <span class="va">orthologs</span><span class="op">$</span><span class="va">GeneID</span><span class="op">)</span>
<span class="va">xx</span> <span class="op">=</span> <span class="va">orthologs</span><span class="op">$</span><span class="va">HumanSymbol</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span>
<span class="va">logik</span> <span class="op">=</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">orthologs</span><span class="op">$</span><span class="va">HumanSymbol</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span><span class="op">)</span>  <span class="co"># sum(logik) returns 17,365 homologous genes</span>
<span class="va">xx</span> <span class="op">=</span> <span class="va">xx</span><span class="op">[</span><span class="va">logik</span><span class="op">]</span>
<span class="va">orthologs</span> <span class="op">=</span> <span class="va">orthologs</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">orthologs</span><span class="op">$</span><span class="va">HumanSymbol</span><span class="op">)</span>, <span class="op">]</span>
<span class="co"># note: several of these genes are not unique mappers; we will aggregate them later or make them</span>
<span class="co"># unique. To aggregate, where E is the sparse expression matrix with rownames set to xx: E =</span>
<span class="co"># Matrix.utils::aggregate.Matrix(E, row.names(E))</span></code></pre></div>
<p>Now we have mapped homologous gene symbols across species:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">orthologs</span><span class="op">)</span></code></pre></div>
<pre><code>##               GeneID cynoSymbol HumanSymbol     HumanGeneID
## 1 ENSMFAG00000046426        ND6      MT-ND6 ENSG00000198695
## 2 ENSMFAG00000002805      POLG2       POLG2 ENSG00000256525
## 3 ENSMFAG00000046418       COX2      MT-CO2 ENSG00000198712
## 4 ENSMFAG00000042657    SLC38A3     SLC38A3 ENSG00000188338
## 5 ENSMFAG00000042891      HMOX1       HMOX1 ENSG00000100292
## 6 ENSMFAG00000038079     SHISA9      SHISA9 ENSG00000237515</code></pre>
<p>After mapping homologous genes, Signac can be used to classify the cell types.</p>
<details><summary><strong>Session Info</strong>
</summary><pre><code>## R version 4.0.0 (2020-04-24)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: CentOS Linux 7 (Core)
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib64/libopenblasp-r0.3.3.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## loaded via a namespace (and not attached):
##  [1] knitr_1.31        magrittr_2.0.1    R6_2.5.0          ragg_1.1.1       
##  [5] rlang_0.4.10      fastmap_1.1.0     stringr_1.4.0     tools_4.0.0      
##  [9] xfun_0.21         jquerylib_0.1.3   htmltools_0.5.1.1 systemfonts_1.0.1
## [13] yaml_2.2.1        assertthat_0.2.1  digest_0.6.27     rprojroot_2.0.2  
## [17] pkgdown_1.6.1     crayon_1.4.1      textshaping_0.3.1 formatR_1.7      
## [21] sass_0.3.1        fs_1.5.0          memoise_2.0.0     cachem_1.0.3     
## [25] evaluate_0.14     rmarkdown_2.7     stringi_1.5.3     compiler_4.0.0   
## [29] bslib_0.2.4       desc_1.2.0        jsonlite_1.7.2</code></pre>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Mathew Chamberlain, Virginia Savova, Richa Hanamsagar, Frank Nestle, Emanuele de Rinaldis, Sanofi US.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
