<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>SignacX, Seurat and MASC: Analysis of kidney cells from AMP • SignacX</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="SignacX, Seurat and MASC: Analysis of kidney cells from AMP">
<meta property="og:description" content="SignacX">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SignacX</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/install.html">Install</a>
</li>
<li>
  <a href="../articles/dataportal.html">Data portal</a>
</li>
<li>
  <a href="../articles/quickstart.html">Quick start</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Introductory Vignettes</li>
    <li>
      <a href="../articles/signac-Seurat_pbmcs.html">PBMCs 1K tutorial - Seurat integration</a>
    </li>
    <li>
      <a href="../articles/springintegration.html">Integrating SignacX with SPRING and Seurat</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Advanced use</li>
    <li>
      <a href="../articles/signac-SPRING_Learning.html">Learning from single cell data</a>
    </li>
    <li>
      <a href="../articles/Crabeating_vignette.html">Classifying non-human single cell data</a>
    </li>
    <li>
      <a href="../articles/signac-Seurat_AMP.html">Identifying disease-enriched cell types with SignacX</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Benchmarking</li>
    <li>
      <a href="../articles/signac-Seurat_AMP_RA.html">Flow cytometry synovium: SingleR vs. SignacX</a>
    </li>
    <li>
      <a href="../articles/SignacFast-Seurat_AMP_RA.html">Flow cytometry synovium: FastSignac vs Signac</a>
    </li>
    <li>
      <a href="../articles/signac-Seurat_CITE-seq.html">CITE-seq PBMCs</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Other</li>
    <li>
      <a href="../articles/genesofinterest.html">Genes of immune and pharmacological interest</a>
    </li>
  </ul>
</li>
<li>
  <a href="https://github.com/mathewchamberlain/SignacX/discussions">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mathewchamberlain/SignacX/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="signac-Seurat_AMP_files/header-attrs-2.7/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>SignacX, Seurat and MASC: Analysis of kidney cells from AMP</h1>
            
            <h4 class="date">Compiled 2021-03-02</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/mathewchamberlain/SignacX/blob/master/vignettes/signac-Seurat_AMP.Rmd"><code>vignettes/signac-Seurat_AMP.Rmd</code></a></small>
      <div class="hidden name"><code>signac-Seurat_AMP.Rmd</code></div>

    </div>

    
    
<p>Sometimes, we have single cell genomics data with disease information, and we want to know which cellular phenotypes are enriched for disease. In this vignette, we applied SignacX to classify cellular phenotypes in healthy and lupus nephritis kidney cells, and then we used <a href="https://pubmed.ncbi.nlm.nih.gov/30333237/">MASC</a> to identify which cellular phenotypes were disease-enriched.</p>
<p>Note:</p>
<ul>
<li>MASC typically requires equal numbers of cells and samples between case and control: an unequal number might skew the clustering of cells towards one sample (i.e., a “batch effect”), which could cause spurious disease enrichment in the mixed effect model. Since Signac classifies each cell independently (without using clusters), Signac annotations can be used with MASC without a priori balancing samples or cells, unlike cluster-based annotation methods.</li>
</ul>
<p>This vignette shows how to use SignacX with Seurat and MASC. There are three parts: Seurat, SignacX and then MASC. We use an example data set of kidney cells from AMP from <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6726437/">this publication</a>.</p>
<div id="load-data" class="section level2">
<h2 class="hasAnchor">
<a href="#load-data" class="anchor"></a>Load data</h2>
<p>Read the CEL-seq2 data.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ReadCelseq</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">counts.file</span>, <span class="va">meta.file</span><span class="op">)</span> <span class="op">{</span>
    <span class="va">E</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html">suppressWarnings</a></span><span class="op">(</span><span class="fu">readr</span><span class="fu">::</span><span class="fu">read_tsv</span><span class="op">(</span><span class="va">counts.file</span><span class="op">)</span><span class="op">)</span>
    <span class="va">gns</span> <span class="op">&lt;-</span> <span class="va">E</span><span class="op">$</span><span class="va">gene</span>
    <span class="va">E</span> <span class="op">=</span> <span class="va">E</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span>
    <span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html">suppressWarnings</a></span><span class="op">(</span><span class="fu">readr</span><span class="fu">::</span><span class="fu">read_tsv</span><span class="op">(</span><span class="va">meta.file</span><span class="op">)</span><span class="op">)</span>
    <span class="va">S</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">M</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span>
        <span class="va">logik</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">E</span><span class="op">)</span> <span class="op">%in%</span> <span class="va">M</span><span class="op">$</span><span class="va">cell_name</span><span class="op">[</span><span class="va">M</span><span class="op">$</span><span class="va">sample</span> <span class="op">==</span> <span class="va">x</span><span class="op">]</span>
        <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/Matrix.html">Matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">E</span><span class="op">[</span>, <span class="va">logik</span><span class="op">]</span><span class="op">)</span>, sparse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
    <span class="op">}</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">S</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">M</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span>
    <span class="va">S</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">S</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span>
        <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">gns</span>
        <span class="va">x</span>
    <span class="op">}</span><span class="op">)</span>
    <span class="va">S</span>
<span class="op">}</span>

<span class="va">counts.file</span> <span class="op">=</span> <span class="st">"./SDY997_EXP15176_celseq_matrix_ru10_molecules.tsv.gz"</span>
<span class="va">meta.file</span> <span class="op">=</span> <span class="st">"./SDY997_EXP15176_celseq_meta.tsv"</span>

<span class="va">E</span> <span class="op">=</span> <span class="fu">ReadCelseq</span><span class="op">(</span>counts.file <span class="op">=</span> <span class="va">counts.file</span>, meta.file <span class="op">=</span> <span class="va">meta.file</span><span class="op">)</span>
<span class="va">M</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html">suppressWarnings</a></span><span class="op">(</span><span class="fu">readr</span><span class="fu">::</span><span class="fu">read_tsv</span><span class="op">(</span><span class="va">meta.file</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Merge into a single matrix</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># keep only kidney cells</span>
<span class="va">E</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">E</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span>
    <span class="va">x</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">"K"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>
<span class="op">}</span><span class="op">)</span>

<span class="co"># remove any sample with no cells</span>
<span class="va">E</span> <span class="op">=</span> <span class="va">E</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">E</span>, <span class="va">ncol</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">]</span>

<span class="co"># merge into a single matrix</span>
<span class="va">E</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="va">cbind</span>, <span class="va">E</span><span class="op">)</span></code></pre></div>
</div>
<div id="seurat" class="section level2">
<h2 class="hasAnchor">
<a href="#seurat" class="anchor"></a>Seurat</h2>
<p>Start with the standard pre-processing steps for a Seurat object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></code></pre></div>
<p>Create a Seurat object, and then perform SCTransform normalization. Note:</p>
<ul>
<li>You can use the legacy functions here (i.e., NormalizeData, ScaleData, etc.), use SCTransform or any other normalization method (including no normalization). We did not notice a significant difference in cell type annotations with different normalization methods.</li>
<li>We think that it is best practice to use SCTransform, but it is not a necessary step. Signac will work fine without it.</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load data</span>
<span class="va">kidney</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/CreateSeuratObject.html">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">E</span>, project <span class="op">=</span> <span class="st">"celseq"</span><span class="op">)</span>

<span class="co"># run sctransform</span>
<span class="va">kidney</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html">SCTransform</a></span><span class="op">(</span><span class="va">kidney</span><span class="op">)</span></code></pre></div>
<p>Perform dimensionality reduction by PCA and UMAP embedding. Note:</p>
<ul>
<li>SignacX actually needs these functions since it uses the nearest neighbor graph generated by Seurat.</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># These are now standard steps in the Seurat workflow for visualization and clustering</span>
<span class="va">kidney</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html">RunPCA</a></span><span class="op">(</span><span class="va">kidney</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">kidney</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">kidney</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">kidney</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span><span class="va">kidney</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
</div>
<div id="signacx" class="section level2">
<h2 class="hasAnchor">
<a href="#signacx" class="anchor"></a>SignacX</h2>
<p>First, make sure that you have the SignacX package installed.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"SignacX"</span><span class="op">)</span></code></pre></div>
<p>Generate cellular phenotype labels for the Seurat object. Note:</p>
<ul>
<li>Optionally, you can do parallel computing by setting num.cores &gt; 1 in the Signac function.</li>
<li>Run time is ~10 minutes for ~10,000 cells on a single core.</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Run Signac</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mathewchamberlain/SignacX">SignacX</a></span><span class="op">)</span>
<span class="va">labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Signac.html">Signac</a></span><span class="op">(</span><span class="va">kidney</span>, num.cores <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>
<span class="va">celltypes</span> <span class="op">=</span> <span class="fu"><a href="../reference/GenerateLabels.html">GenerateLabels</a></span><span class="op">(</span><span class="va">labels</span>, E <span class="op">=</span> <span class="va">kidney</span><span class="op">)</span></code></pre></div>
</div>
<div id="visualizations" class="section level2">
<h2 class="hasAnchor">
<a href="#visualizations" class="anchor"></a>Visualizations</h2>
<p>Now we can visualize the cell type classifications at many different levels:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kidney</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/AddMetaData.html">AddMetaData</a></span><span class="op">(</span><span class="va">kidney</span>, metadata <span class="op">=</span> <span class="va">celltypes</span><span class="op">$</span><span class="va">Immune</span>, col.name <span class="op">=</span> <span class="st">"immmune"</span><span class="op">)</span>
<span class="va">kidney</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Idents.html">SetIdent</a></span><span class="op">(</span><span class="va">kidney</span>, value <span class="op">=</span> <span class="st">"immmune"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/png.html">png</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="st">"fls/plot1_amp.png"</span><span class="op">)</span>
<span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">kidney</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="fls/plot1_amp.png" alt=""><p class="caption">Immune, Nonimmune (if any) and unclassified cells</p>
</div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kidney</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/AddMetaData.html">AddMetaData</a></span><span class="op">(</span><span class="va">kidney</span>, metadata <span class="op">=</span> <span class="va">celltypes</span><span class="op">$</span><span class="va">CellTypes</span>, col.name <span class="op">=</span> <span class="st">"celltypes"</span><span class="op">)</span>
<span class="va">kidney</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Idents.html">SetIdent</a></span><span class="op">(</span><span class="va">kidney</span>, value <span class="op">=</span> <span class="st">"celltypes"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/png.html">png</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="st">"fls/plot2_amp.png"</span><span class="op">)</span>
<span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">kidney</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="fls/plot2_amp.png" alt=""><p class="caption">Cell types</p>
</div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kidney</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/AddMetaData.html">AddMetaData</a></span><span class="op">(</span><span class="va">kidney</span>, metadata <span class="op">=</span> <span class="va">celltypes</span><span class="op">$</span><span class="va">CellStates</span>, col.name <span class="op">=</span> <span class="st">"cellstates"</span><span class="op">)</span>
<span class="va">kidney</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Idents.html">SetIdent</a></span><span class="op">(</span><span class="va">kidney</span>, value <span class="op">=</span> <span class="st">"cellstates"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/png.html">png</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="st">"fls/plot3_amp.png"</span><span class="op">)</span>
<span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">kidney</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="fls/plot3_amp.png" alt=""><p class="caption">Cell states</p>
</div>
<p>Identify immune marker genes (IMAGES)</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Downsample just the immune cells</span>
<span class="va">kidney.small</span> <span class="op">&lt;-</span> <span class="va">kidney</span><span class="op">[</span>, <span class="op">!</span><span class="va">celltypes</span><span class="op">$</span><span class="va">CellStates</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"NonImmune"</span>, <span class="st">"Fibroblasts"</span>, <span class="st">"Unclassified"</span>, 
    <span class="st">"Endothelial"</span>, <span class="st">"Epithelial"</span><span class="op">)</span><span class="op">]</span>

<span class="co"># Find protein markers for all clusters, and draw a heatmap</span>
<span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers</a></span><span class="op">(</span><span class="va">kidney.small</span>, only.pos <span class="op">=</span> <span class="cn">TRUE</span>, verbose <span class="op">=</span> <span class="cn">F</span>, logfc.threshold <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="va">top5</span> <span class="op">&lt;-</span> <span class="va">markers</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html">top_n</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">5</span>, wt <span class="op">=</span> <span class="va">avg_logFC</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/png.html">png</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="st">"fls/plot4_amp.png"</span>, width <span class="op">=</span> <span class="fl">640</span>, height <span class="op">=</span> <span class="fl">640</span><span class="op">)</span>
<span class="fu"><a href="https://satijalab.org/seurat/reference/DoHeatmap.html">DoHeatmap</a></span><span class="op">(</span><span class="va">kidney.small</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">top5</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span>, angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="fls/plot4_amp.png" alt=""><p class="caption">Immune marker genes</p>
</div>
</div>
<div id="run-masc" class="section level2">
<h2 class="hasAnchor">
<a href="#run-masc" class="anchor"></a>Run MASC</h2>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Meta_mapped</span> <span class="op">=</span> <span class="va">M</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">kidney</span><span class="op">)</span>, <span class="va">M</span><span class="op">$</span><span class="va">cell_name</span><span class="op">)</span>, <span class="op">]</span>
<span class="va">Meta_mapped</span><span class="op">$</span><span class="va">CellStates</span> <span class="op">=</span> <span class="va">celltypes</span><span class="op">$</span><span class="va">CellStates</span>
<span class="va">Meta_mapped</span><span class="op">$</span><span class="va">disease</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">Meta_mapped</span><span class="op">$</span><span class="va">disease</span><span class="op">)</span>
<span class="va">Q</span> <span class="op">=</span> <span class="fu"><a href="../reference/MASC.html">MASC</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="va">Meta_mapped</span>, cluster <span class="op">=</span> <span class="va">Meta_mapped</span><span class="op">$</span><span class="va">CellStates</span>, contrast <span class="op">=</span> <span class="st">"disease"</span>, random_effects <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"plate"</span>, 
    <span class="st">"lane"</span>, <span class="st">"sample"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>MASC results reveals that fibroblasts, plasma cells and B memory cells are enriched (p value &lt; 0.05) for disease.</p>
<table class="table">
<thead><tr>
<th style="text-align:left;">
cluster
</th>
<th style="text-align:right;">
size
</th>
<th style="text-align:right;">
model.pvalue
</th>
<th style="text-align:right;">
diseaseSLE.OR
</th>
<th style="text-align:right;">
diseaseSLE.OR.95pct.ci.lower
</th>
<th style="text-align:right;">
diseaseSLE.OR.95pct.ci.upper
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
B.memory
</td>
<td style="text-align:right;">
391
</td>
<td style="text-align:right;">
0.0197539
</td>
<td style="text-align:right;">
3.041576e+00
</td>
<td style="text-align:right;">
1.1841285
</td>
<td style="text-align:right;">
7.812654e+00
</td>
</tr>
<tr>
<td style="text-align:left;">
B.naive
</td>
<td style="text-align:right;">
144
</td>
<td style="text-align:right;">
0.4652596
</td>
<td style="text-align:right;">
1.548526e+00
</td>
<td style="text-align:right;">
0.4932181
</td>
<td style="text-align:right;">
4.861810e+00
</td>
</tr>
<tr>
<td style="text-align:left;">
DC
</td>
<td style="text-align:right;">
160
</td>
<td style="text-align:right;">
0.0536208
</td>
<td style="text-align:right;">
2.346694e+00
</td>
<td style="text-align:right;">
1.0278135
</td>
<td style="text-align:right;">
5.357951e+00
</td>
</tr>
<tr>
<td style="text-align:left;">
Endothelial
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
0.2290332
</td>
<td style="text-align:right;">
1.183979e+01
</td>
<td style="text-align:right;">
0.1021348
</td>
<td style="text-align:right;">
1.372508e+03
</td>
</tr>
<tr>
<td style="text-align:left;">
Epithelial
</td>
<td style="text-align:right;">
89
</td>
<td style="text-align:right;">
0.5615659
</td>
<td style="text-align:right;">
6.424482e-01
</td>
<td style="text-align:right;">
0.1488724
</td>
<td style="text-align:right;">
2.772440e+00
</td>
</tr>
<tr>
<td style="text-align:left;">
Fibroblasts
</td>
<td style="text-align:right;">
360
</td>
<td style="text-align:right;">
0.0174935
</td>
<td style="text-align:right;">
3.926459e+00
</td>
<td style="text-align:right;">
2.5047560
</td>
<td style="text-align:right;">
6.155124e+00
</td>
</tr>
<tr>
<td style="text-align:left;">
Macrophages
</td>
<td style="text-align:right;">
356
</td>
<td style="text-align:right;">
0.1061625
</td>
<td style="text-align:right;">
4.700437e-01
</td>
<td style="text-align:right;">
0.1959415
</td>
<td style="text-align:right;">
1.127587e+00
</td>
</tr>
<tr>
<td style="text-align:left;">
Mon.Classical
</td>
<td style="text-align:right;">
247
</td>
<td style="text-align:right;">
0.6346884
</td>
<td style="text-align:right;">
8.482458e-01
</td>
<td style="text-align:right;">
0.4416789
</td>
<td style="text-align:right;">
1.629059e+00
</td>
</tr>
<tr>
<td style="text-align:left;">
Mon.NonClassical
</td>
<td style="text-align:right;">
658
</td>
<td style="text-align:right;">
0.7360915
</td>
<td style="text-align:right;">
8.693916e-01
</td>
<td style="text-align:right;">
0.3903675
</td>
<td style="text-align:right;">
1.936231e+00
</td>
</tr>
<tr>
<td style="text-align:left;">
Neutrophils
</td>
<td style="text-align:right;">
11
</td>
<td style="text-align:right;">
0.3806773
</td>
<td style="text-align:right;">
1.800080e-02
</td>
<td style="text-align:right;">
0.0000020
</td>
<td style="text-align:right;">
1.590775e+02
</td>
</tr>
<tr>
<td style="text-align:left;">
NK
</td>
<td style="text-align:right;">
562
</td>
<td style="text-align:right;">
0.2814722
</td>
<td style="text-align:right;">
6.046977e-01
</td>
<td style="text-align:right;">
0.2470403
</td>
<td style="text-align:right;">
1.480160e+00
</td>
</tr>
<tr>
<td style="text-align:left;">
NonImmune
</td>
<td style="text-align:right;">
112
</td>
<td style="text-align:right;">
0.1608263
</td>
<td style="text-align:right;">
7.817140e-02
</td>
<td style="text-align:right;">
0.0021302
</td>
<td style="text-align:right;">
2.868614e+00
</td>
</tr>
<tr>
<td style="text-align:left;">
Plasma.cells
</td>
<td style="text-align:right;">
93
</td>
<td style="text-align:right;">
0.0055213
</td>
<td style="text-align:right;">
5.387766e+06
</td>
<td style="text-align:right;">
0.0000000
</td>
<td style="text-align:right;">
1.579379e+25
</td>
</tr>
<tr>
<td style="text-align:left;">
T.CD4.memory
</td>
<td style="text-align:right;">
283
</td>
<td style="text-align:right;">
0.5351536
</td>
<td style="text-align:right;">
6.933100e-01
</td>
<td style="text-align:right;">
0.2284183
</td>
<td style="text-align:right;">
2.104379e+00
</td>
</tr>
<tr>
<td style="text-align:left;">
T.CD4.naive
</td>
<td style="text-align:right;">
459
</td>
<td style="text-align:right;">
0.4627315
</td>
<td style="text-align:right;">
7.488830e-01
</td>
<td style="text-align:right;">
0.3546322
</td>
<td style="text-align:right;">
1.581429e+00
</td>
</tr>
<tr>
<td style="text-align:left;">
T.CD8.cm
</td>
<td style="text-align:right;">
482
</td>
<td style="text-align:right;">
0.1776758
</td>
<td style="text-align:right;">
1.917053e+00
</td>
<td style="text-align:right;">
0.7686826
</td>
<td style="text-align:right;">
4.781025e+00
</td>
</tr>
<tr>
<td style="text-align:left;">
T.CD8.em
</td>
<td style="text-align:right;">
1196
</td>
<td style="text-align:right;">
0.7985566
</td>
<td style="text-align:right;">
1.107881e+00
</td>
<td style="text-align:right;">
0.5097435
</td>
<td style="text-align:right;">
2.407877e+00
</td>
</tr>
<tr>
<td style="text-align:left;">
T.CD8.naive
</td>
<td style="text-align:right;">
39
</td>
<td style="text-align:right;">
0.8117720
</td>
<td style="text-align:right;">
1.132623e+00
</td>
<td style="text-align:right;">
0.4022482
</td>
<td style="text-align:right;">
3.189161e+00
</td>
</tr>
<tr>
<td style="text-align:left;">
T.regs
</td>
<td style="text-align:right;">
84
</td>
<td style="text-align:right;">
0.6186262
</td>
<td style="text-align:right;">
1.571148e+00
</td>
<td style="text-align:right;">
0.2607223
</td>
<td style="text-align:right;">
9.467957e+00
</td>
</tr>
<tr>
<td style="text-align:left;">
Unclassified
</td>
<td style="text-align:right;">
717
</td>
<td style="text-align:right;">
0.8900750
</td>
<td style="text-align:right;">
9.198574e-01
</td>
<td style="text-align:right;">
0.2840429
</td>
<td style="text-align:right;">
2.978908e+00
</td>
</tr>
</tbody>
</table>
<p>Save results</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">kidney</span>, file <span class="op">=</span> <span class="st">"fls/amp_kidney_signac.rds"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">Q</span>, file <span class="op">=</span> <span class="st">"fls/amp_kidney_MASC_result.rds"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">celltypes</span>, file <span class="op">=</span> <span class="st">"fls/amp_kidney_celltypes.rds"</span><span class="op">)</span></code></pre></div>
<details><summary><strong>Session Info</strong>
</summary><pre><code>## R version 4.0.0 (2020-04-24)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: CentOS Linux 7 (Core)
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib64/libopenblasp-r0.3.3.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## loaded via a namespace (and not attached):
##  [1] knitr_1.31        magrittr_2.0.1    R6_2.5.0          ragg_1.1.1       
##  [5] rlang_0.4.10      fastmap_1.1.0     highr_0.8         stringr_1.4.0    
##  [9] tools_4.0.0       xfun_0.21         jquerylib_0.1.3   htmltools_0.5.1.1
## [13] systemfonts_1.0.1 yaml_2.2.1        assertthat_0.2.1  digest_0.6.27    
## [17] rprojroot_2.0.2   pkgdown_1.6.1     crayon_1.4.1      textshaping_0.3.1
## [21] formatR_1.7       sass_0.3.1        fs_1.5.0          memoise_2.0.0    
## [25] cachem_1.0.3      evaluate_0.14     rmarkdown_2.7     stringi_1.5.3    
## [29] compiler_4.0.0    bslib_0.2.4       desc_1.2.0        jsonlite_1.7.2</code></pre>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Mathew Chamberlain, Virginia Savova, Richa Hanamsagar, Frank Nestle, Emanuele de Rinaldis, Sanofi US.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
