<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Signac and SPRING: Learning CD56 NK cells from multi-modal analysis of CITE-seq PBMCs from 10X Genomics • SignacX</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Signac and SPRING: Learning CD56 NK cells from multi-modal analysis of CITE-seq PBMCs from 10X Genomics">
<meta property="og:description" content="SignacX">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SignacX</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Crabeating_vignette.html">Mapping homologous gene symbols</a>
    </li>
    <li>
      <a href="../articles/SignacFast-Seurat_AMP_RA.html">Benchmarking SignacFast with single cell flow-sorted data</a>
    </li>
    <li>
      <a href="../articles/signac-SPRING_Learning.html">Signac and SPRING: Learning CD56 NK cells from multi-modal analysis of CITE-seq PBMCs from 10X Genomics</a>
    </li>
    <li>
      <a href="../articles/signac-Seurat_AMP.html">SignacX, Seurat and MASC: Analysis of kidney cells from AMP</a>
    </li>
    <li>
      <a href="../articles/signac-Seurat_AMP_RA.html">Benchmarking SignacX and SingleR with synovial flow cytometry data</a>
    </li>
    <li>
      <a href="../articles/signac-Seurat_CITE-seq.html">Cell type classification with SignacX: CITE-seq PBMCs from 10X Genomics</a>
    </li>
    <li>
      <a href="../articles/signac-Seurat_pbmcs.html">Cell type classification with SignacX: PBMCs from 10X Genomics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mathewchamberlain/SignacX/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="signac-SPRING_Learning_files/header-attrs-2.7/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Signac and SPRING: Learning CD56 NK cells from multi-modal analysis of CITE-seq PBMCs from 10X Genomics</h1>
                        <h4 class="author">Mathew Chamberlain</h4>
            
            <h4 class="date">2021-03-01</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/mathewchamberlain/SignacX/blob/master/vignettes/signac-SPRING_Learning.Rmd"><code>vignettes/signac-SPRING_Learning.Rmd</code></a></small>
      <div class="hidden name"><code>signac-SPRING_Learning.Rmd</code></div>

    </div>

    
    
<p>This vignette shows how to use SignacX with Seurat and SPRING to learn a new cell type category from single cell data.</p>
<div id="load-data" class="section level2">
<h2 class="hasAnchor">
<a href="#load-data" class="anchor"></a>Load data</h2>
<p>We start with CITE-seq data that were already classified with SignacX using the SPRING pipeline.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mathewchamberlain/SignacX">SignacX</a></span><span class="op">)</span></code></pre></div>
<p>Load CITE-seq data from 10X Genomics processed with SPRING and classified with SignacX already.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load CITE-seq data</span>
<span class="va">data.dir</span> <span class="op">=</span> <span class="st">'./CITESEQ_EXPLORATORY_CITESEQ_5K_PBMCS/FullDataset_v1_protein'</span>
<span class="va">E</span> <span class="op">=</span> <span class="fu"><a href="../reference/CID.LoadData.html">CID.LoadData</a></span><span class="op">(</span>data.dir <span class="op">=</span> <span class="va">data.dir</span><span class="op">)</span>

<span class="co"># Load labels</span>
<span class="va">json_data</span> <span class="op">=</span> <span class="fu">rjson</span><span class="fu">::</span><span class="fu">fromJSON</span><span class="op">(</span>file<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">data.dir</span>,<span class="st">'/categorical_coloring_data.json'</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Create a Seurat object for the protein expression data; we will use this as a reference.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># separate protein and gene expression data</span>
<span class="va">logik</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">"Total"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">E</span><span class="op">)</span><span class="op">)</span>
<span class="va">P</span> <span class="op">=</span> <span class="va">E</span><span class="op">[</span><span class="va">logik</span>,<span class="op">]</span>
<span class="va">E</span> <span class="op">=</span> <span class="va">E</span><span class="op">[</span><span class="op">!</span><span class="va">logik</span>,<span class="op">]</span>

<span class="co"># CLR normalization in Seurat</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">P</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">P</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">E</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">E</span><span class="op">)</span>
<span class="va">reference</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/CreateSeuratObject.html">CreateSeuratObject</a></span><span class="op">(</span><span class="va">E</span><span class="op">)</span>
<span class="va">reference</span><span class="op">[[</span><span class="st">"ADT"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/CreateAssayObject.html">CreateAssayObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">P</span><span class="op">)</span>
<span class="va">reference</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html">NormalizeData</a></span><span class="op">(</span><span class="va">reference</span>, assay <span class="op">=</span> <span class="st">"ADT"</span>, normalization.method <span class="op">=</span> <span class="st">"CLR"</span><span class="op">)</span></code></pre></div>
<p>Identify CD56 bright NK cells based on protein expression data.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># generate labels </span>
<span class="va">lbls</span> <span class="op">=</span> <span class="va">json_data</span><span class="op">$</span><span class="va">CellStates</span><span class="op">$</span><span class="va">label_list</span>
<span class="va">lbls</span><span class="op">[</span><span class="va">lbls</span> <span class="op">!=</span> <span class="st">"NK"</span><span class="op">]</span> <span class="op">=</span> <span class="st">"Unclassified"</span>
<span class="va">CD16</span> <span class="op">=</span> <span class="va">reference</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">ADT</span><span class="op">@</span><span class="va">counts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">reference</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">ADT</span><span class="op">@</span><span class="va">counts</span><span class="op">)</span> <span class="op">==</span> <span class="st">"CD16-TotalSeqB-CD16"</span>,<span class="op">]</span>
<span class="va">CD56</span> <span class="op">=</span> <span class="va">reference</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">ADT</span><span class="op">@</span><span class="va">counts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">reference</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">ADT</span><span class="op">@</span><span class="va">counts</span><span class="op">)</span> <span class="op">==</span> <span class="st">"CD56-TotalSeqB-CD56"</span>,<span class="op">]</span>
<span class="va">logik</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">CD56</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">10</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">CD16</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">7.5</span> <span class="op">&amp;</span> <span class="va">lbls</span> <span class="op">==</span> <span class="st">"NK"</span>; <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">logik</span><span class="op">)</span>
<span class="va">lbls</span><span class="op">[</span><span class="va">logik</span><span class="op">]</span> <span class="op">=</span> <span class="st">"NK.CD56bright"</span></code></pre></div>
</div>
<div id="signacx" class="section level2">
<h2 class="hasAnchor">
<a href="#signacx" class="anchor"></a>SignacX</h2>
<p>Generate a training data set from the reference data and save it for later use. Note:</p>
<ul>
<li>SignacBoot performs feature selection, bootstrapping, imputation and normalization to derive a training data set from single cell data.</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># generate bootstrapped single cell data</span>
<span class="va">R_learned</span> <span class="op">=</span> <span class="fu"><a href="../reference/SignacBoot.html">SignacBoot</a></span><span class="op">(</span>E <span class="op">=</span> <span class="va">E</span>, spring.dir <span class="op">=</span> <span class="va">data.dir</span>, L <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"NK"</span>, <span class="st">"NK.CD56bright"</span><span class="op">)</span>, labels <span class="op">=</span> <span class="va">lbls</span>, logfc.threshold <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>

<span class="co"># save the training data</span>
<span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span><span class="op">(</span><span class="va">R_learned</span>, file <span class="op">=</span> <span class="st">"training_NKBright_v207.rda"</span><span class="op">)</span></code></pre></div>
</div>
<div id="classify-a-new-data-set-with-the-model" class="section level2">
<h2 class="hasAnchor">
<a href="#classify-a-new-data-set-with-the-model" class="anchor"></a>Classify a new data set with the model</h2>
<p>Load expression data for a different data set (this was also previously processed through SPRING and SignacX)</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Classify another data set with new model</span>
<span class="co"># load new data</span>
<span class="va">new.data.dir</span> <span class="op">=</span> <span class="st">"./PBMCs_5k_10X/FullDataset_v1"</span>
<span class="va">E</span> <span class="op">=</span> <span class="fu"><a href="../reference/CID.LoadData.html">CID.LoadData</a></span><span class="op">(</span>data.dir <span class="op">=</span> <span class="va">new.data.dir</span><span class="op">)</span>
<span class="co"># load cell types identified with Signac</span>
<span class="va">json_data</span> <span class="op">=</span> <span class="fu">rjson</span><span class="fu">::</span><span class="fu">fromJSON</span><span class="op">(</span>file<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">new.data.dir</span>,<span class="st">'/categorical_coloring_data.json'</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Generate new labels. Note:</p>
<ul>
<li>Signac trains an ensemble of 100 neural network classifiers using the new training data set built above (R_learned), and then classifies unseen data (E).</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># generate new labels</span>
<span class="va">cr_learned</span> <span class="op">=</span> <span class="fu"><a href="../reference/Signac.html">Signac</a></span><span class="op">(</span>E <span class="op">=</span> <span class="va">E</span>, R <span class="op">=</span> <span class="va">R_learned</span>, spring.dir <span class="op">=</span> <span class="va">new.data.dir</span><span class="op">)</span></code></pre></div>
<p>Now we amend the existing labels (classified previously with SignacX); we add the new labels and generate a new SPRING layout.Note:</p>
<ul>
<li>We usually copy the existing SPRING files from “FullDataset_v1” to “FullDataset_v1_Learned” to generate a new layout while preserving the existing layout.</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># modify the existing labels</span>
<span class="va">cr</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">json_data</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">$</span><span class="va">label_list</span><span class="op">)</span>
<span class="va">logik</span> <span class="op">=</span> <span class="va">cr</span><span class="op">$</span><span class="va">CellStates</span> <span class="op">==</span> <span class="st">'NK'</span>
<span class="va">cr</span><span class="op">$</span><span class="va">CellStates</span><span class="op">[</span><span class="va">logik</span><span class="op">]</span> <span class="op">=</span> <span class="va">cr_learned</span><span class="op">[</span><span class="va">logik</span><span class="op">]</span>
<span class="va">logik</span> <span class="op">=</span> <span class="va">cr</span><span class="op">$</span><span class="va">CellStates_novel</span> <span class="op">==</span> <span class="st">'NK'</span>
<span class="va">cr</span><span class="op">$</span><span class="va">CellStates_novel</span><span class="op">[</span><span class="va">logik</span><span class="op">]</span> <span class="op">=</span> <span class="va">cr_learned</span><span class="op">[</span><span class="va">logik</span><span class="op">]</span>
<span class="va">new.data.dir</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">new.data.dir</span>, <span class="st">"_Learned"</span><span class="op">)</span></code></pre></div>
<p>Save results</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># save</span>
<span class="va">dat</span> <span class="op">=</span> <span class="fu"><a href="../reference/CID.writeJSON.html">CID.writeJSON</a></span><span class="op">(</span><span class="va">cr</span>, spring.dir <span class="op">=</span> <span class="va">new.data.dir</span>, new_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'red'</span><span class="op">)</span>, new_populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span> <span class="st">'NK.CD56bright'</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<details><summary><strong>Session Info</strong>
</summary><pre><code>## R version 4.0.0 (2020-04-24)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: CentOS Linux 7 (Core)
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib64/libopenblasp-r0.3.3.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## loaded via a namespace (and not attached):
##  [1] rprojroot_2.0.2   crayon_1.4.1      digest_0.6.27     assertthat_0.2.1 
##  [5] R6_2.5.0          jsonlite_1.7.2    magrittr_2.0.1    evaluate_0.14    
##  [9] stringi_1.5.3     rlang_0.4.10      cachem_1.0.3      fs_1.5.0         
## [13] jquerylib_0.1.3   bslib_0.2.4       ragg_1.1.1        rmarkdown_2.7    
## [17] pkgdown_1.6.1     textshaping_0.3.1 desc_1.2.0        tools_4.0.0      
## [21] stringr_1.4.0     yaml_2.2.1        xfun_0.21         fastmap_1.1.0    
## [25] compiler_4.0.0    systemfonts_1.0.1 memoise_2.0.0     htmltools_0.5.1.1
## [29] knitr_1.31        sass_0.3.1</code></pre>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Mathew Chamberlain, Virginia Savova, Richa Hanamsagar, Frank Nestle, Emanuele de Rinaldis, Sanofi US.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
