<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Fast classification of cellular phenotypes — SignacFast • SignacX</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Fast classification of cellular phenotypes — SignacFast" />
<meta property="og:description" content="SignacFast uses pre-computed neural network models to classify cellular phenotypes in single cell data:
these models were pre-trained with the HPCA training data. Any features that are
present in the training data and absent in the single cell data are set to zero. 
This is a factor of ~5-10 speed improvement over Signac." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SignacX</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Crabeating_vignette.html">Mapping homologous gene symbols</a>
    </li>
    <li>
      <a href="../articles/SignacFast-Seurat_AMP_RA.html">Benchmarking SignacFast with single cell flow-sorted data</a>
    </li>
    <li>
      <a href="../articles/signac-SPRING_Learning.html">Signac and SPRING: Learning CD56 NK cells from multi-modal analysis of CITE-seq PBMCs from 10X Genomics</a>
    </li>
    <li>
      <a href="../articles/signac-Seurat_AMP.html">SignacX, Seurat and MASC: Analysis of kidney cells from AMP</a>
    </li>
    <li>
      <a href="../articles/signac-Seurat_AMP_RA.html">Benchmarking SignacX and SingleR with synovial flow cytometry data</a>
    </li>
    <li>
      <a href="../articles/signac-Seurat_CITE-seq.html">Cell type classification with SignacX: CITE-seq PBMCs from 10X Genomics</a>
    </li>
    <li>
      <a href="../articles/signac-Seurat_pbmcs.html">Cell type classification with SignacX: PBMCs from 10X Genomics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/mathewchamberlain/SignacX/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Fast classification of cellular phenotypes</h1>
    <small class="dont-index">Source: <a href='https://github.com/mathewchamberlain/SignacX/blob/master/R/SignacFast.R'><code>R/SignacFast.R</code></a></small>
    <div class="hidden name"><code>SignacFast.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>SignacFast</code> uses pre-computed neural network models to classify cellular phenotypes in single cell data:
these models were pre-trained with the HPCA training data. Any features that are
present in the training data and absent in the single cell data are set to zero. 
This is a factor of ~5-10 speed improvement over <code><a href='Signac.html'>Signac</a></code>.</p>
    </div>

    <pre class="usage"><span class='fu'>SignacFast</span><span class='op'>(</span>
  <span class='va'>E</span>,
  Models <span class='op'>=</span> <span class='st'>"default"</span>,
  spring.dir <span class='op'>=</span> <span class='cn'>NULL</span>,
  num.cores <span class='op'>=</span> <span class='fl'>1</span>,
  threshold <span class='op'>=</span> <span class='fl'>0</span>,
  smooth <span class='op'>=</span> <span class='cn'>TRUE</span>,
  impute <span class='op'>=</span> <span class='cn'>TRUE</span>,
  verbose <span class='op'>=</span> <span class='cn'>TRUE</span>,
  do.normalize <span class='op'>=</span> <span class='cn'>TRUE</span>,
  return.probability <span class='op'>=</span> <span class='cn'>FALSE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>E</th>
      <td><p>a gene (rows) by cell (column) matrix, sparse matrix or a Seurat object. Rows are HUGO symbols.</p></td>
    </tr>
    <tr>
      <th>Models</th>
      <td><p>if 'default', as returned by <code><a href='GetModels_HPCA.html'>GetModels_HPCA</a></code>. An ensemble of 1,800 neural network models.</p></td>
    </tr>
    <tr>
      <th>spring.dir</th>
      <td><p>If using SPRING, directory to categorical_coloring_data.json. Default is NULL.</p></td>
    </tr>
    <tr>
      <th>num.cores</th>
      <td><p>number of cores to use for parallel computation. Default is 1.</p></td>
    </tr>
    <tr>
      <th>threshold</th>
      <td><p>Probability threshold for assigning cells to "Unclassified." Default is 0.</p></td>
    </tr>
    <tr>
      <th>smooth</th>
      <td><p>if TRUE, smooths the cell type classifications. Default is TRUE.</p></td>
    </tr>
    <tr>
      <th>impute</th>
      <td><p>if TRUE, gene expression values are imputed prior to cell type classification (see <code><a href='KSoftImpute.html'>KSoftImpute</a></code>). Default is TRUE.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>if TRUE, code will report outputs. Default is TRUE.</p></td>
    </tr>
    <tr>
      <th>do.normalize</th>
      <td><p>if TRUE, cells are normalized to the mean library size. Default is TRUE.</p></td>
    </tr>
    <tr>
      <th>return.probability</th>
      <td><p>if TRUE, returns the probability associated with each cell type label. Default is TRUE.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list of character vectors: cell type annotations (L1, L2, ...) at each level of the hierarchy
as well as 'clusters' for the Louvain clustering results.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='Signac.html'>Signac</a></code> for another classification function.</p>
<p><code><a href='Signac.html'>Signac</a></code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>
<span class='co'># download single cell data for classification</span>
<span class='va'>file.dir</span> <span class='op'>=</span> <span class='st'>"https://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_1k_v3/"</span>
<span class='va'>file</span> <span class='op'>=</span> <span class='st'>"pbmc_1k_v3_filtered_feature_bc_matrix.h5"</span>
<span class='fu'><a href='https://rdrr.io/r/utils/download.file.html'>download.file</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='va'>file.dir</span>, <span class='va'>file</span><span class='op'>)</span>, <span class='st'>"Ex.h5"</span><span class='op'>)</span>

<span class='co'># load data, process with Seurat</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://satijalab.org/seurat'>Seurat</a></span><span class='op'>)</span>
<span class='va'>E</span> <span class='op'>=</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/Read10X_h5.html'>Read10X_h5</a></span><span class='op'>(</span>filename <span class='op'>=</span> <span class='st'>"Ex.h5"</span><span class='op'>)</span>
<span class='va'>pbmc</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/CreateSeuratObject.html'>CreateSeuratObject</a></span><span class='op'>(</span>counts <span class='op'>=</span> <span class='va'>E</span>, project <span class='op'>=</span> <span class='st'>"pbmc"</span><span class='op'>)</span>
<span class='va'>pbmc</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/SCTransform.html'>SCTransform</a></span><span class='op'>(</span><span class='va'>pbmc</span><span class='op'>)</span>
<span class='va'>pbmc</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/RunPCA.html'>RunPCA</a></span><span class='op'>(</span><span class='va'>pbmc</span>, verbose <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
<span class='va'>pbmc</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/RunUMAP.html'>RunUMAP</a></span><span class='op'>(</span><span class='va'>pbmc</span>, dims <span class='op'>=</span> <span class='fl'>1</span><span class='op'>:</span><span class='fl'>30</span>, verbose <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
<span class='va'>pbmc</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/FindNeighbors.html'>FindNeighbors</a></span><span class='op'>(</span><span class='va'>pbmc</span>, dims <span class='op'>=</span> <span class='fl'>1</span><span class='op'>:</span><span class='fl'>30</span>, verbose <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>

<span class='co'># classify cells</span>
<span class='va'>labels</span> <span class='op'>=</span> <span class='fu'>SignacFast</span><span class='op'>(</span>E <span class='op'>=</span> <span class='va'>pbmc</span><span class='op'>)</span>
<span class='va'>celltypes</span> <span class='op'>=</span> <span class='fu'><a href='GenerateLabels.html'>GenerateLabels</a></span><span class='op'>(</span><span class='va'>labels</span>, E <span class='op'>=</span> <span class='va'>pbmc</span><span class='op'>)</span>

<span class='co'># add labels to Seurat object, visualize</span>
<span class='va'>lbls</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span><span class='va'>celltypes</span><span class='op'>$</span><span class='va'>CellStates</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='va'>lbls</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/sort.html'>sort</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/unique.html'>unique</a></span><span class='op'>(</span><span class='va'>lbls</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>pbmc</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/AddMetaData.html'>AddMetaData</a></span><span class='op'>(</span><span class='va'>pbmc</span>, metadata<span class='op'>=</span><span class='va'>celltypes</span><span class='op'>$</span><span class='va'>CellStates</span>, col.name <span class='op'>=</span> <span class='st'>"celltypes"</span><span class='op'>)</span>
<span class='va'>pbmc</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://satijalab.org/seurat/reference/Idents.html'>SetIdent</a></span><span class='op'>(</span><span class='va'>pbmc</span>, value<span class='op'>=</span><span class='st'>'celltypes'</span><span class='op'>)</span>
<span class='fu'><a href='https://satijalab.org/seurat/reference/DimPlot.html'>DimPlot</a></span><span class='op'>(</span><span class='va'>pbmc</span>, label <span class='op'>=</span> <span class='cn'>T</span><span class='op'>)</span>

<span class='co'># save results</span>
<span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>saveRDS</a></span><span class='op'>(</span><span class='va'>pbmc</span>, <span class='st'>"pbmcs.rds"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>saveRDS</a></span><span class='op'>(</span><span class='va'>celltypes</span>, <span class='st'>"celltypes.rds"</span><span class='op'>)</span>
<span class='op'>}</span>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Mathew Chamberlain, Virginia Savova, Richa Hanamsagar, Frank Nestle, Emanuele de Rinaldis, Sanofi US.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


